Source/WebCore/platform/graphics/RenderingResource.h

```diff
@@ -44,23 +44,43 @@ class RenderingResource
 
     virtual ~RenderingResource()
     {
+        if (!hasValidRenderingResourceIdentifier())
+            return;
         for (auto observer : m_observers)
-            observer->releaseRenderingResource(m_renderingResourceIdentifier);
+            observer->releaseRenderingResource(renderingResourceIdentifier());
     }
 
-    RenderingResourceIdentifier renderingResourceIdentifier() const { return m_renderingResourceIdentifier; }
+    bool hasValidRenderingResourceIdentifier() const
+    {
+        return m_renderingResourceIdentifier.has_value();
+    }
```

Comment:
When do we create resources without a valid `RenderingResourceIdentifier`? Is it that we don't provide an identifier for gradients that are cheap to send across or something? If so, instead of doing it based on whether the identifier is valid, is it something the encode/decode functions can decide by inspecting the gradient? Then we can continue to have valid identifiers always.

Reply:
When trying to cache all gradients, I noticed MotionMark CanvasLines regressed significantly. The reason for this is CanvasLines creates a gradient every time a line is drawn https://github.com/WebKit/WebKit/blob/main/PerformanceTests/MotionMark/tests/master/resources/canvas-tests.js#L238. So caching this gradient was not beneficial at all since each gradient is just used only once. This contrary to SVG gradients which lives as long as the SVG object lives. So only SVG gradients will have valid `RenderingResourceIdentifier`.